### PR TITLE
Move Editions Index page to the GOV.UK Design System

### PR DESCRIPTION
**Pull request by davidgisbey**
## Description

This adds the editions index page in the GOV.UK Design system. It's turned into quite a chunky PR, mainly just due to the page being quite large.

I've tried to break it into smaller commits as best i can so it's less noisy to review.

This follows the established pattern of prepending `legacy` the Bootstrap implementation, then adding the GOV.UK Design System one.

I've had a to tweak the `SelectWithSearch` Component and the `PaginationHelper` to fix a couple of bugs/issues.

The `SelectWithSearch` doesn't have the option to pass in a blank option without it being overwritten/set as a placeholder. I've tweaked it so that if a option is passed in with text, but no value it doesn't overwrite the text with `Select one` and have overwritten the placeholder so that it has black text.

There might be a better long term solution for this. I'll chat to Ollie about it when he's back.

I've also fixed an issue with pagination, where it wasn't passing through query string values when another page was clicked as it was using the path, not the url.

## Screenshots

image

When published state is selected

image

With no pagination

image

With no results 

image

## Trello card

https://trello.com/c/Fe2Ei59d/18-document-listings-search

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

### CODE REVIEW

**Comment by nnagewad** [comment]
Looking good. Design nitpicks:

- Have spacing between results and black horizontal rule should be
    - `15px` below returned results
    -  `20px` below the black section break
- Returned results number should use: `govuk-heading-s`

**Comment by davidgisbey** [comment]
Thanks., What do we do for the force published stats? All bold?

**Comment by chriswhitehouse-tw** [comment]
Some initial manual user testing observations:

* When a user selects an option, the re-render is not triggered until the Search button is also clicked. On the original implementation, the user selection triggered the re-render.
* The pagination component wraps round to the next line, when page 2 is selected: 
Screenshot 2023-04-20 at 12 12 51

File: app/controllers/admin/editions_controller.rb [hunk 1]
```diff
@@ -417,13 +417,16 @@ def filter
...
-    params_filters_with_default_state
+    filter_options = params_filters_with_default_state
       .symbolize_keys
       .merge(
         include_unpublishing: true,
         include_link_check_reports: true,
         include_last_author: true,
       )
+
+    filter_options = filter_options.merge(per_page: 15) if preview_design_system?(next_release: false)
```
Comment (chriswhitehouse-tw) [inline app/controllers/admin/editions_controller.rb:428]: should the 15 be made a constant somewhere rather than hardcoded here?

**Review by chriswhitehouse-tw — COMMENTED** [review]
Hey David, I've had a review and left a few specific comments.

Broadly looks good, and seems to work as intended based on manual tests.

One comment on the commit history, should the 'fixup feature specs' commit be tidied up into the prior implementation commits?

Happy/keen to discuss, if you any Qs.

File: app/views/admin/editions/_search_results.html.erb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        text: "Title"
+      },
+      {
+        text: "Updated by"
+      },
+      {
+        text: "State"
+      },
+      {
+        text: tag.span("State", class: "govuk-visually-hidden")
```
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/views/admin/editions/_search_results.html.erb:24]: oh good catch! It definitely should

File: app/controllers/admin/editions_controller.rb [hunk 3]
```diff
@@ -195,6 +195,7 @@ def history
   def get_layout
     design_system_actions = %w[confirm_destroy diff show]
     design_system_actions += %w[edit update new create] if preview_design_system?(next_release: true)
+    design_system_actions += %w[index] if preview_design_system?(next_release: false)
```
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/controllers/admin/editions_controller.rb:198]: Yea that's exactly it 👍

File: app/controllers/admin/editions_controller.rb (hunk 1 above)
Comment (davidgisbey) [reply app/controllers/admin/editions_controller.rb:428]: yea i think you're right. It'll likely be used across multiple pages so it makes sense to.

**Comment by davidgisbey** [comment]
```
- When a user selects an option, the re-render is not triggered until the Search button is also clicked. On the original implementation, the user selection triggered the re-render.
- The pagination component wraps round to the next line, when page 2 is selected:
```

1. Good spot. We discussed this and decided not to replicate this functionality for now for a couple of reasons (complexity and consistency as it's only place used in Whitehall). We may circle back to it later

File: app/models/admin/edition_filter.rb [hunk 4]
```diff
@@ -3,6 +3,7 @@ class EditionFilter
     EDITION_TYPE_LOOKUP = Whitehall.edition_classes.index_by(&:to_s)
 
     MAX_EXPORT_SIZE = 8000
+    GOVUK_DESIGN_SYSTEM_PER_PAGE = 15
```
Comment (chriswhitehouse-tw) [inline app/models/admin/edition_filter.rb:6]: This seems an appropriate spot for the constant to be stored.

File: app/views/admin/editions/_search_results.html.erb [hunk 5]
```diff
@@ -21,7 +21,7 @@
         text: "State"
       },
       {
-        text: tag.span("State", class: "govuk-visually-hidden")
+        text: tag.span("Edit", class: "govuk-visually-hidden")
```
Comment (chriswhitehouse-tw) [inline app/views/admin/editions/_search_results.html.erb:24]: cool

File: app/views/admin/editions/_search_results.html.erb [hunk 6]
```diff
@@ -48,10 +48,10 @@
 
     <% if paginate(paginator, theme: "govuk_paginator").present? && show_export && can?(:export, Edition) %>
       <div class="govuk-grid-row">
-        <div class="govuk-grid-column-two-thirds">
+        <div class="govuk-grid-column-three-quarters">
```
Comment (chriswhitehouse-tw) [inline app/views/admin/editions/_search_results.html.erb:51]: Nice, yes, this seems to work 👍

**Review by chriswhitehouse-tw — APPROVED** [review]
LGTM!

**Pull request merged**